js/scorm/wrapper.js

```diff
@@ -592,8 +593,12 @@ class ScormWrapper {
       validate(`${cmiPrefix}.score.max`, maxScore);
     }
     this.setValue(`${cmiPrefix}.score.raw`, score);
-    if (this.isSupported(`${cmiPrefix}.score.min`)) this.setValue(`${cmiPrefix}.score.min`, minScore);
-    if (this.isSupported(`${cmiPrefix}.score.max`)) this.setValue(`${cmiPrefix}.score.max`, maxScore);
+    const children = this.getValue(`${cmiPrefix}.score._children`);
+    if (this.isUnsupportedLastError()) return;
+    children.includes('min') ? this.setValue(`${cmiPrefix}.score.min`, minScore)
+      : this.logUnsupported(`${cmiPrefix}.score.min`);
+    children.includes('max') ? this.setValue(`${cmiPrefix}.score.max`, maxScore)
+      : this.logUnsupported(`${cmiPrefix}.score.max`);
```

Comment:
We tend not to use ternaries for conditional statement execution but instead as expressions for generating a value from conditions.

There is a lot of repeating behaviour here which you're trying to reduce with ternaries.

Perhaps add `isChildSupported`?:

```js
if (this.isChildSupported(`${cmiPrefix}.score.min`)) this.setValue(`${cmiPrefix}.score.min`, minScore);

isChildSupported(propertyPath) {
  const propertyPathParts = propertyPath.split('.');
  const childPropertyName = propertyPathParts.pop();
  const parentPropertyPath = propertyPathParts.join('.');
  const children = this.getValue(`${parentPropertyPath}._children`);
  if (this.isUnsupportedLastError()) {
    this.logUnsupported(propertyPath);
    return false;
  }
  return children.includes(childPropertyName);
}
```

Reply:
> this could result in lots of additional API calls.

With SCORM the API call will only look up sub properties from a nested state object, it will not be a resource hungry process, caching the children variable will not give any noticeable performance improvement, optimisation through variable caching is completely unnecessary here and is causing obtuse code. We're not talking about a slow API, such as one that will cross javascript contexts across Window boundaries or perform HTTP requests, read files from disk etc, this is just traversing local javascript object properties with a few checks. SCORM's getValue and setValue are almost identical to Backbone's model set and get, except that they allow sub properties through dot name paths and have a few additional checks.

[Moodle scorm 1.2 API implementation](https://github.com/moodle/moodle/blob/9587029a4609b3222a3564207aada02716d3332d/mod/scorm/datamodels/scorm_12.js#L242-L406)

Reply:
I understand how the API works and that the SCO isn't making HTTP requests, but I wasn't clear how little the performance would be impacted by making multiple calls compared to one. I thought calling the API `LMSGetValue`/`GetValue` to lookup the models multiple times for the same method would be inefficient. Can we be sure that all LMSs implement the API calls efficiently?

I was also mindful that the SCORM log becomes noisier as a result, making it more obtuse when being used to analyse diagnostics from SCORM Test Packages.

Ultimately, if we are happy to lean on the API more to reduce code, without impacting performance, I can adjust accordingly.

Reply:
It's always a cached state object. Sometimes it's a nested hierarchy, sometimes a flat name value hash like on Moodle / Totara. The API is synchronous for get and set and commit and finish are used to push the states to the lms. 

The scorm logging provided by the wrapper in the function `this.getValue`, for a function like `this.isChildSupported`, can be circumvented by using `this.scorm.get` directly. As it is in `this.isSupported` currently. Checking for support in a new function wouldn't be logged as it isn't logged currently.

The performance impact of the optimization in this code would always only be client side and with a high degree of certainty, be always completely negligible. We set suspendData with greater frequency than we'll ever set interactions.

Reply:
`isSupported` does log calls - https://github.com/adaptlearning/adapt-contrib-spoor/blob/master/js/scorm/wrapper.js#L471, but I get your point that it directly targets the pipwerks library, so doesn't add to the log for the get value requests.

Good point regarding `suspend_data` and `lesson_location`/`location` requests which are constantly firing without issue.

Reply:
Updated support logic to include unsupported logging in the methods checking support:
https://github.com/adaptlearning/adapt-contrib-spoor/pull/306/commits/0c6836b508df040989a4e693191684137a1970f0

Adding supported checks into `setValue` and returning early if not supported would have been nice and only needed for SCORM 1.2, but would ideally include some form of conditional checking against acceptable property values to ensure we aren't triggering 201 LMS errors for values which don't have `_children`. I didn't really want to include hardcoded references to capture the following models which allow write access to their children:
- cmi.core.score
- cmi.objectives
- cmi.objectives.n.score
- cmi.student_preference
- cmi.interactions